base/logging.jl

```diff
@@ -268,75 +276,27 @@ default_group(file) = Symbol(splitext(basename(file))[1])
 
 # Generate code for logging macros
 function logmsg_code(_module, file, line, level, message, exs...)
-    id = Expr(:quote, log_record_id(_module, level, message, exs))
-    group = nothing
-    kwargs = Any[]
-    for ex in exs
-        if ex isa Expr && ex.head === :(=) && ex.args[1] isa Symbol
-            k,v = ex.args
-            if !(k isa Symbol)
-                throw(ArgumentError("Expected symbol for key in key value pair `$ex`"))
-            end
-            k = ex.args[1]
-            # Recognize several special keyword arguments
-            if k === :_id
-                # id may be overridden if you really want several log
-                # statements to share the same id (eg, several pertaining to
-                # the same progress step).  In those cases it may be wise to
-                # manually call log_record_id to get a unique id in the same
-                # format.
-                id = esc(v)
-            elseif k === :_module
-                _module = esc(v)
-            elseif k === :_line
-                line = esc(v)
-            elseif k === :_file
-                file = esc(v)
-            elseif k === :_group
-                group = esc(v)
-            else
-                # Copy across key value pairs for structured log records
-                push!(kwargs, Expr(:kw, k, esc(v)))
-            end
-        elseif ex isa Expr && ex.head === :...
-            # Keyword splatting
-            push!(kwargs, esc(ex))
-        else
-            # Positional arguments - will be converted to key value pairs
-            # automatically.
-            push!(kwargs, Expr(:kw, Symbol(ex), esc(ex)))
-        end
-    end
-
-    if group === nothing
-        group = if isdefined(Base, :basename) && isa(file, String)
-            # precompute if we can
-            QuoteNode(default_group(file))
-        else
-            # memoized run-time execution
-            ref = Ref{Symbol}()
-            :(isassigned($ref) ? $ref[]
-                               : $ref[] = default_group(something($file, "")))
-        end
-    end
-
+    log_data = process_logmsg_exs(_module, file, line, level, message, exs...)
```

Comment:
I think it is nicer to have 1 log_data object to hold all the data about the log, other than the level and message.
(tempting to put them in there too).